### PR TITLE
Support for flux exec scripts

### LINKED ISSUE #3114: Flux executor needs to respect `named_env`, `pre_exec` etc
Probably should use a similar shell script as the POPEN executor

### PR DESCRIPTION
**Pull request by andre-merzky**
This fixes #3114

### CODE REVIEW

**Comment by mtitov** [comment]
@andre-merzky I've tested it on Frontier and got the following errors (probably `flux.dump` is not crucial, but the one in `agent_execution` is):

```
$ cat flux.0039.log 
1710449568.901 : flux.0039            : 19460 : 140737179921280 : DEBUG    : flux output: URI:local:///tmp/flux-I1uzqO/local-0
1710449568.903 : flux.0039            : 19460 : 140731208460032 : INFO     : starting flux watcher
1710449568.903 : flux.0039            : 19460 : 140737179921280 : INFO     : flux startup successful: [ssh://frontier07386.frontier.olcf.ornl.gov/tmp/flux-I1uzqO/local-0]
```

```
$ cat flux.0040.dump 
connect flux 19424: ssh://frontier07386.frontier.olcf.ornl.gov/tmp/flux-I1uzqO/local-0
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/bin/radical-pilot-agent_0", line 17, in <module>
    agent = rp.Agent_0()
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/agent_0.py", line 47, in __init__
    self._session = Session(uid=cfg.sid, cfg=cfg, _role=Session._AGENT_0)
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/session.py", line 192, in __init__
    elif self._role == self._AGENT_0: self._init_agent_0()
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/session.py", line 288, in _init_agent_0
    self._init_rm()
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/session.py", line 812, in _init_rm
    self._rm = ResourceManager.create(name=rname,
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/resource_manager/base.py", line 382, in create
    return rm(cfg, rcfg, log, prof)
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/resource_manager/base.py", line 161, in __init__
    self._prepare_launch_methods()
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/resource_manager/base.py", line 339, in _prepare_launch_methods
    self._launchers[lm_name] = rpa.LaunchMethod.create(
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/launch_method/base.py", line 148, in create
    return impl[name](name, lm_cfg, rm_info, log, prof)
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/launch_method/flux.py", line 18, in __init__
    LaunchMethod.__init__(self, name, lm_cfg, rm_info, session, prof)
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/launch_method/base.py", line 91, in __init__
    self._init_from_info(lm_info)
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/launch_method/flux.py", line 62, in _init_from_info
    self._fh.connect_flux(uri=self._details['flux_uri'])
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/utils/flux.py", line 419, in connect_flux
    for l in get_stacktrace():
```

```
$ cat agent_executing.0000.log 
...
1710449602.816 : agent_executing.0000 : 20215 : 140735505663744 : DEBUG    : submit tasks: [{'tasks': [{'slot': 'task', 'count': {'per_slot': 1}, 'command': ['/bin/sh', '-c', '/lustre/orion/scratch/matitov/chm155/radical.pilot.sandbox/rp.session.frontier07386.matitov.019796.0000/pilot.0000/task.000000//task.000000.exec.sh 1>/lustre/orion/scratch/matitov/chm155/radical.pilot.sandbox/rp.session.frontier07386.matitov.019796.0000/pilot.0000/task.000000//task.000000.out 2>/lustre/orion/scratch/matitov/chm155/radical.pilot.sandbox/rp.session.frontier07386.matitov.019796.0000/pilot.0000/task.000000//task.000000.err']}], 'attributes': {'system': {'cwd': '/lustre/orion/scratch/matitov/chm155/radical.pilot.sandbox/rp.session.frontier07386.matitov.019796.0000/pilot.0000/task.000000/', 'duration': 0}}, 'version': 1, 'resources': [{'count': 8, 'type': 'slot', 'label': 'task', 'with': [{'count': 1, 'type': 'core'}, {'count': 1.0, 'type': 'gpu'}]}]}]
1710449603.099 : agent_executing.0000 : 20215 : 140735505663744 : ERROR    : work <bound method Flux.work of <radical.pilot.agent.executing.flux.Flux object at 0x7fffe4cd9d00>> failed
Traceback (most recent call last):
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/utils/component.py", line 1032, in work_cb
    self._workers[state](things)
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/pilot/agent/executing/flux.py", line 129, in work
    jids = self._lm.fh.submit_jobs([jd for jd in jds])
  File "/lustre/orion/chm155/scratch/matitov/flux/ve.rp/lib/python3.9/site-packages/radical/utils/flux.py", line 477, in submit_jobs
    flux_id = fut.get_id()
  File "/usr/lib64/python3.6/site-packages/flux/job/submit.py", line 26, in get_id
    return submit_get_id(self)
  File "/usr/lib64/python3.6/site-packages/flux/util.py", line 70, in func_wrapper
    raise EnvironmentError(error.errno, errmsg.decode("utf-8")) from None
PermissionError: [Errno 1] count must be an int or mapping
1710449603.100 : agent_executing.0000 : 20215 : 140735505663744 : DEBUG    : advance bulk: 1 [False, True, FAILED]
```

(*) single task with 8 ranks, 1 core per rank and 1 gpu per rank, and it has its own conda env (which is activated in its pre-exec section)

**Comment by andre-merzky** [comment]
> (*) single task with 8 ranks, 1 core per rank and 1 gpu per rank, and it has its own conda env (which is activated in its pre-exec section)

Thanks @mtitov - this was caused by our float type for `gpus_per_rank` and should be fixed now - please do give it another try.

File: src/radical/pilot/agent/scheduler/continuous.py [hunk 1]
```diff
@@ -230,7 +230,7 @@ def _find_resources(self, node, find_slots, ranks_per_slot, cores_per_slot,
             while len(gpus) < gpus_per_slot:
 
                 if node['gpus'][gpu_idx] == rpc.FREE:
-                    gpus.append(gpu_idx)
+                gpus.append(gpu_idx)
```
Comment (mtitov) [inline src/radical/pilot/agent/scheduler/continuous.py:233]: typo

File: src/radical/pilot/agent/executing/flux.py [hunk 2]
```diff
@@ -188,16 +188,22 @@ def task_to_spec(self, task):
...
-                    'count': td['gpus_per_rank'],
-                    'type' : 'gpu'})
+
+             gpr = td['gpus_per_rank']
+
+             if gpr != int(gpr):
+                 raise ValueError('flux does not support on-integer GPU count')
+
+             spec['resources'][0]['with'].append({
+                     'count': gpr,
```
Comment (mtitov) [inline src/radical/pilot/agent/executing/flux.py:205]: `'count': int(gpr),`

**Review by mtitov — COMMENTED** [review]
I've got the following logs in `agent_executing` (execution went through, but task failed)

```
 DEBUG    : submit tasks: [{'tasks': [{'slot': 'task', 'count': {'per_slot': 1}, 'command': ['/bin/sh', '-c', '/lustre/orion/scratch/matitov/chm155/flux/radical.pilot.sandbox/rp.session.frontier06474.matitov.019797.0008/pilot.0000/task.000000//task.000000.exec.sh 1>/lustre/orion/scratch/matitov/chm155/flux/radical.pilot.sandbox/rp.session.frontier06474.matitov.019797.0008/pilot.0000/task.000000//task.000000.out 2>/lustre/orion/scratch/matitov/chm155/flux/radical.pilot.sandbox/rp.session.frontier06474.matitov.019797.0008/pilot.0000/task.000000//task.000000.err']}], 'attributes': {'system': {'cwd': '/lustre/orion/scratch/matitov/chm155/flux/radical.pilot.sandbox/rp.session.frontier06474.matitov.019797.0008/pilot.0000/task.000000/', 'duration': 0}}, 'version': 1, 'resources': [{'count': 8, 'type': 'slot', 'label': 'task', 'with': [{'count': 1, 'type': 'core'}, {'count': 1, 'type': 'gpu'}]}]}]
 DEBUG    : submitted tasks
 DEBUG    : submitted task task.000000 -> fFyAASJj
 DEBUG    : handle task.000000: fFyAASJj
 DEBUG    : ignore flux event task.000000:submit
 DEBUG    : task state: None [1710517156.33022: submit {'userid': 16142, 'urgency': 16, 'flags': 0, 'version': 1}]
 DEBUG    : advance bulk: 1 [False, True, AGENT_EXECUTING]
 DEBUG    : task state: FAILED [1710517156.35185: exception {'type': 'alloc', 'severity': 0, 'note': "Unsupported resource type 'gpu'"}]
 DEBUG    : advance bulk: 1 [False, True, FAILED]
```

File: src/radical/pilot/agent/executing/flux.py (hunk 2 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/executing/flux.py:205]: oops - thanks!

File: src/radical/pilot/agent/scheduler/continuous.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/scheduler/continuous.py:233]: Thanks!

**Comment by andre-merzky** [comment]
> Is this related to [flux-framework/flux-core#3372](https://github.com/flux-framework/flux-core/discussions/3372)?

I don't think so - we don't use `sched-simple`.  Maybe the remaining `float` type was causing this - how does it look now?  Anything in the `flux*.log` files?

**Comment by mtitov** [comment]
seems that Flux doesn't see GPUs
```
flux resources [ 0 ]:
     STATE NNODES   NCORES    NGPUS NODELIST
      free      1       56        0 frontier00178
 allocated      0        0        0 
      down      0        0        0 

```

(*) attached flux-logs
[flux.logs.tar.gz](https://github.com/radical-cybertools/radical.pilot/files/14639860/flux.logs.tar.gz)

p.s. I've made this run also using flux update in RU - it works!

**Comment by andre-merzky** [comment]
I guess we first need to fix the flux startup then - the missing GPUs might be caused by not using `srun` to start flux?  dunno...

File: src/radical/pilot/agent/executing/dragon.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        super().initialize()
+
+        self._pid = self.session.cfg.pid
+
+        # run dragon execution helper in a separate process which actually uses
+        # dragon's multiprocessing implementation
+        self._dragon = sp.Popen(
+                args   = ['python3', # 'dragon',
+                          '/home/merzky/j/rp.2/bin/radical-pilot-dragon-executor.py',
```
Comment (mtitov) [inline src/radical/pilot/agent/executing/dragon.py:42]: hardcoded path

**Review by mtitov — APPROVED** [review]
LGTM!
(Dragon part wasn't tested, but there is a comment on hardcoded path)

p.s. flux doesn't see GPUs no matter how it was started, that what we need to check with OLCF support, but it doesn't block this PR from merging

File: src/radical/pilot/agent/executing/dragon.py (hunk 3 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/executing/dragon.py]: Oops - fixed!

**Pull request merged**